bootstrap-salt.sh

```diff
@@ -7004,26 +7004,37 @@ install_suse_check_services() {
 #    Gentoo Install Functions.
 #
 __autounmask() {
-    emerge --autounmask-write --autounmask-only "${@}"; return $?
+    # Unmask package(s) and accept changes
+    #
+    # Usually it's a good thing to have config files protected by portage, but
+    # in this case this would require to interrupt the bootstrapping script at
+    # this point, manually merge the changes using etc-update/dispatch-conf/
+    # cfg-update and then restart the bootstrapping script, so instead we allow
+    # at this point to modify certain config files directly
+    export CONFIG_PROTECT_MASK="${CONFIG_PROTECT_MASK:-}
+        /etc/portage/package.accept_keywords
+        /etc/portage/package.keywords
+        /etc/portage/package.license
+        /etc/portage/package.unmask
+        /etc/portage/package.use"
+    emerge --autounmask --autounmask-continue --autounmask-only --autounmask-write "${@}"; return $?
 }
 
 __emerge() {
-    if [ "$_GENTOO_USE_BINHOST" -eq $BS_TRUE ]; then
-        emerge --getbinpkg "${@}"; return $?
+    EMERGE_FLAGS='-q'
+    if [ "$_ECHO_DEBUG" -eq $BS_TRUE ]; then
+        EMERGE_FLAGS='-v'
     fi
-    emerge "${@}"; return $?
-}
 
-__gentoo_config_protection() {
-    # usually it's a good thing to have config files protected by portage, but
-    # in this case this would require to interrupt the bootstrapping script at
-    # this point, manually merge the changes using etc-update/dispatch-conf/
-    # cfg-update and then restart the bootstrapping script, so instead we allow
-    # at this point to modify certain config files directly
-    export CONFIG_PROTECT_MASK="${CONFIG_PROTECT_MASK:-} /etc/portage/package.accept_keywords /etc/portage/package.keywords /etc/portage/package.license /etc/portage/package.unmask /etc/portage/package.use"
+    # Do not re-emerge packages that are already installed
+    EMERGE_FLAGS="${EMERGE_FLAGS} --noreplace"
+
+    if [ "$_GENTOO_USE_BINHOST" -eq $BS_TRUE ]; then
+        EMERGE_FLAGS="${EMERGE_FLAGS} --getbinpkg"
+    fi
 
-    # emerge currently won't write to files that aren't there, so we need to ensure their presence
-    touch /etc/portage/package.accept_keywords /etc/portage/package.keywords /etc/portage/package.license /etc/portage/package.unmask /etc/portage/package.use
+    # shellcheck disable=SC2086
+    emerge ${EMERGE_FLAGS} "${@}"; return $?
 }
 
 __gentoo_pre_dep() {
```

Comment:
Several tests failed due to the 20 minutes per build limit. Some time could be saved by using emerge-webrsync instead of emerge --sync, but that may not be sufficient.

Reply:
Should we add a `emerge-webrsync` to the kitchen config file for Gentoo?
At least its up to date when the test container is created....

Reply:
I like the idea of adding `emerge-webrsync` to the `provision_command` section, It is in the PR now.

Unfortunately it does not solve the problem. I went through the logs of one of the successful runs and it turned out that kitchen-docker runs `emerge --sync` when creating the container - see https://github.com/test-kitchen/kitchen-docker/blob/24a54be8dece9d095e0f9285c509def1c186f088/lib/kitchen/docker/helpers/dockerfile_helper.rb#L81-L88 

I guess that the solution is to replace `emerge --sync` by `emerge-webrsync` in kitchen-docker. I will play with it and open a PR against that code to see if it can be replaced or not.

Reply:
With the changes in test-kitchen/kitchen-docker#373, more Gentoo builds succeed, but some of them still get cancelled. I guess that the only two options are either to skip running Gentoo tests in the pipeline or raise the 20 minutes build limit. What do you think, @s0undt3ch?

Reply:
We actually need to wait for the kitchen-docker release or install from git on the Gemfile

Reply:
I updated the Gemfile to pull the latest kitchen-docker master in https://github.com/saltstack/salt-bootstrap/pull/1500/commits/821cc2a792a129f3a1ff14d46e6d127d8915344d. The last build was executed with these changes applied.

Reply:
Think 30 minutes will do?

Reply:
@s0undt3ch These figures for building the pre-salted images from the bootstrap may be in the same sort of ballpark:

* 15 minutes 7 seconds: [`git-master-py3`](https://gitlab.com/myii/salt-image-builder/-/jobs/759275102)
* 35 minutes 4 seconds: [`stable-3001-py3`](https://gitlab.com/myii/salt-image-builder/-/jobs/759214114)
* 35 minutes 51 seconds: [`stable-3000.3-py3`](https://gitlab.com/myii/salt-image-builder/-/jobs/759214115)

In fact, the respective equivalent to the first one took 13m 55s for [this PR](https://github.com/saltstack/salt-bootstrap/pull/1500/checks?check_run_id=1176492078).

Reply:
Hmm, let's bump it to 40 minutes then :cry:

Reply:
I am afraid that 45 minutes is closer to what these builds need.
- GIT and Salt >= 3000 ~ 17 minutes
- GIT and Salt 2019 ~ 40-45 minute
- Stable and Salt >= 3000 ~ 35-40 minutes
- Stable and Salt 2019 ~ 4045 minutes

We could drop 2019, since this one takes the most time - it needs to build Python 3.6.

Reply:
Can't drop 2019, it's still supported...

Reply:
@s0undt3ch I was told by various SaltStack employees that the support is only until `Phase 3 support ends` on the following page:

* https://www.saltstack.com/product-support-lifecycle/

So it appears that `2019.2` is only supported for two more days.  Or is there something I'm misssing?

Reply:
I made the job timeout configurable with the default set to 20 minutes unless overridden in a per distro map. Gentoo jobs have the timeout set to 45 minutes at the moment.

Good news is that all jobs succeeded and most Gentoo jobs finished in less than 20 minutes with just a few exceptions. When the support for 2019 gets dropped at some point in the future, the timeout could be decreased to around 30 minutes.